### DISCUSSION TITLE
How to access databases in unit tests?

### DISCUSSION THREAD

**Discussion by AnnikaCodes** [rwf2/Rocket#2051] [Questions]
Hi, I have a struct which wraps a database and provides methods to query it:
```rust
#[cfg_attr(not(test), database("production_db"))]
#[cfg_attr(test, database("test_db"))]
pub struct PostgresConnection(pub postgres::Client);

impl PostgresConnection {
    // ...
}
````

I would like to have access to Rocket's managed instance of this struct outside of request handlers so that I can, for example, use my data-storage methods to add data directly and then test that the API can read and return that data. I could either go for a higher-level testing approach by just testing the API endpoints that add data in the same test, or create a separate database connection and make the crate more abstract. However, I think that Rocket allowing access to its managed databases would be a pretty useful feature - is this something that Rocket can do? If so, how?

Thanks!

**Comment by somehowchris** [comment]
Never use that, but you could use the [`state` method](https://api.rocket.rs/v0.5-rc/rocket/struct.Rocket.html#method.state) on your `rocket()` from your tests.

Let me know if this actually worked or do you have any reference repository I could jump into?
